Add support for running GCI on the GCE cloud provider

andyzheng0831 opened the pull request:
The supports and configs for GCI added here will be used in 1.3 release and upwards. For 1.2.X releases, we will still use cluster/gce/trusty code. Therefore, we will keep maintaining the trusty code as long as 1.2-release branch is not deprecated. Unlike the trusty code, the GCI cluster initialization is through systemd.

This PR is fully tested using e2e tests. I wrote a script to repeatedly spin up a cluster, run CI and slow tests, and then tear down for 15 times for both pure GCI cluster and hybrid cluster (master on GCI, nodes on containervm). The entire tests lasted for more than 3 days, and no failure was shown. I also verified creating trusty cluster, as this PR revises some places affecting it.

@roberthbailey @dchen1107 please review it.

cc/ @fabioy @wonderfly @kubernetes/goog-image FYI.

Commit:
Add support for running GCI on the GCE cloud provider

andyzheng0831 commented:
Update the PR by running "hack/verify-flags-underscore.py -e > hack/verify-flags/exceptions.txt". But I have no idea why there are some flags not related with this PR updated by the script.

roberthbailey commented on build/common.sh:
```diff
@@ -1116,11 +1118,10 @@ function kube::release::gcs::copy_release_artifacts() {
 
   # Having the configure-vm.sh script and trusty code from the GCE cluster
```
should this comment now say "configure-vm.sh script and gci code from ..."

andyzheng0831 replied:
Thanks for catching it. Will correct this.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+#!/bin/bash
+
+# Copyright 2015 The Kubernetes Authors All rights reserved.
```
2016

andyzheng0831 replied:
Will update all the occurrences to 2016.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  # The GCI image has host firewall which drop most inbound/forwarded packets.
+  # We need to add rules to accept all TCP/UDP packets.
+  if iptables -L INPUT | grep "Chain INPUT (policy DROP)" > /dev/null; then
+    echo "Add rules to accpet all inbound TCP/UDP packets"
```
s/accpet/accept

andyzheng0831 replied:
Will update it

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+    iptables -A INPUT -w -p UDP -j ACCEPT
+  fi
+  if iptables -L FORWARD | grep "Chain FORWARD (policy DROP)" > /dev/null; then
+    echo "Add rules to accpet all forwarded TCP/UDP packets"
```
s/accpet/accept

andyzheng0831 replied:
Will update it

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  if [[ ! -e /dev/disk/by-id/google-master-pd ]]; then
+    return
+  fi
+  device_info=$(ls -l /dev/disk/by-id/google-master-pd)
```
can the `ls` command return an error (which will terminate the script)?

andyzheng0831 replied:
"ls" can fail when the given path does not exist. But here we first check if the pd file exists and then "ls" it. It may fail on other things like disk I/O issue (very unusual). When that happens, skipping the "ls" failure is not helpful. So, I think we are safe here.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+    return
+  fi
+  device_info=$(ls -l /dev/disk/by-id/google-master-pd)
+  relative_path=${device_info##* }
```
why the extra space before the closing `}`?

andyzheng0831 replied:
Frankly speaking, I reused the code from ContainerVM's script. It has this extra space https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/configure-vm.sh#l366. @zmerlynn do you know why?

zmerlynn replied:
Ha, this is some funny magic that predates me. http://tldp.org/LDP/abs/html/string-manipulation.html, "Substring removal". This is basically doing a `readlink`.

From my desktop, here's the breakdown:
```
$ ls -l /dev/disk/by-uuid/2162686d-38a6-4808-b061-efd0a1adb9b3
lrwxrwxrwx 1 root root 10 Apr 18 13:16 /dev/disk/by-uuid/2162686d-38a6-4808-b061-efd0a1adb9b3 -> ../../dm-1
$ foo=$(ls -l /dev/disk/by-uuid/2162686d-38a6-4808-b061-efd0a1adb9b3)
$ echo ${foo##* }
../../dm-1
```

But yeah, that's "delete everything matching `" *"`, i.e. the capture up to the last space.

andyzheng0831 replied:
Thanks for the explanation.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  fi
+}
+
+# Finds the master PD device; returns it in MASTER_PD_DEVICE
```
Should we split the master/node configuration scripts into separate files? It would greatly simplify what needs to be loaded onto each node.

andyzheng0831 replied:
The part for master shares many lines of code with the part for node. If we split configure-helper.sh into two files, with one for master and one for node, we will have many duplicate lines in two files, which will bring more maintenance burden. Another way is split configure-helper.sh into three files with one containing the common code. But that is more fragments, and cross-reference reduces readability.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+function mount-master-pd {
+  find-master-pd
+  if [[ -z "${MASTER_PD_DEVICE}" ]]; then
+    return
```
When would we hit this return case?

andyzheng0831 replied:
When env var MASTER_PD_DEVICE is not set. I think it has a problem. If this var is unset, the script will exit due to "set -o nounset". We should check "${MASTER_PD_DEVICE:-}" instead. The ContainerVM's configure-vm.sh has the same problem. I will correct both.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  mkdir -m 700 -p "${mount_point}/var/etcd"
+  ln -s -f "${mount_point}/var/etcd" /var/etcd
+  mkdir -p /etc/srv
+  # Contains the dynamically generated apiserver auth certs and keys
```
nit: End comments with a period.

andyzheng0831 replied:
Will update all occurrences.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  chgrp -R etcd "${mount_point}/var/etcd"
+}
+
+# After the first boot and on upgrade, these files exists on the master-pd
```
s/exists/exist

andyzheng0831 replied:
Corrected

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+      echo "${CA_CERT}" | base64 -d > "${auth_dir}/ca.crt"
+      echo "${MASTER_CERT}" | base64 -d > "${auth_dir}/server.cert"
+      echo "${MASTER_KEY}" | base64 -d > "${auth_dir}/server.key"
+      # Kubecfg cert/key are optional and included for backwards compatibility.
```
You should be able to remove this comment and those last two lines.

andyzheng0831 replied:
We don't need them anymore? I will remove them from here, but leave the ContainerVM's script to @zmerlynn for double check.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+network-name = ${NODE_NETWORK}
+EOF
+  fi
+  if [[ -n "${MULTIZONE:-}" ]]; then
```
I'm not sure we'd hit this case, but if the above if statement is false and this one is true, then the config file won't have [global] at the top which may make it invalid.

andyzheng0831 replied:
You are right. It does not follow the same logic as ContainerVM. The "[global]" part should be moved out of the if block. I will correct it.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  fi
+}
+
+function create-kubeproxy-kubeconfig {
```
Is this still necessary now that kube-proxy runs in a pod?

@mikedanese

mikedanese replied:
Yes this is still neccessary, kube-proxy does not and probably won't ever use in cluster config. It uses the same kubeconfig as it used to.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+
+# Loads kube-system docker images. It is better to do it before starting kubelet,
+# as kubelet will restart docker daemon, which may interfere with loading images.
+function load-docker-images {
```
@mikedanese are we still going to be side-loading these in 1.3?

zmerlynn replied:
Please say no. If you're going to make that change, make it early in a release.

mikedanese replied:
The issue is https://github.com/kubernetes/kubernetes/issues/24123 and is not scheduled for v1.3.

@zmerlynn I'm a bit confused. Did you mean to say, please say yes?

zmerlynn replied:
Er, yes. Please say we're not changing it at this stage in the release (and you confirmed it).

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  local -r img_dir="${KUBE_HOME}/kube-docker-files"
+  if [[ "${KUBERNETES_MASTER:-}" == "true" ]]; then
+    echo "Try to load docker image file kube-apiserver.tar"
+    timeout 30 docker load -i "${img_dir}/kube-apiserver.tar"
```
These may need to be done in a loop in case they fail.

@jlowdermilk

andyzheng0831 replied:
Will update it

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+    flags="${KUBELET_TEST_LOG_LEVEL}"
+  fi
+  if [[ -n "${KUBELET_PORT:-}" ]]; then
+    flags="${flags} --port=${KUBELET_PORT}"
```
This might be easier to read as `flags+=("--port=${KUBELET_PORT}")` like @zmerlynn did in #25437.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  prepare-log-file /var/log/kube-apiserver.log
+
+  # Calculate variables and assemble the command line.
+  local params="--cloud-provider=gce --address=127.0.0.1 --etcd-servers=http://127.0.0.1:4001 --tls-cert-file=/etc/srv/kubernetes/server.cert --tls-private-key-file=/etc/srv/kubernetes/server.key --secure-port=443 --client-ca-file=/etc/srv/kubernetes/ca.crt --token-auth-file=/etc/srv/kubernetes/known_tokens.csv --basic-auth-file=/etc/srv/kubernetes/basic_auth.csv --allow-privileged=true --authori ...
```
Can you format this as one flag / line so that it's easier to read and maintain?

andyzheng0831 replied:
Sure, will update for all long string vars.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+  params="${params} ${log_level}"
+
+  if [[ -n "${PROJECT_ID:-}" && -n "${TOKEN_URL:-}" && -n "${TOKEN_BODY:-}" && -n "${NODE_NETWORK:-}" ]]; then
+    local -r vm_external_ip=$(curl --fail --silent -H 'Metadata-Flavor: Google' "http://metadata/computeMetadata/v1/instance/network-interfaces/0/access-configs/0/external-ip")
```
This curl should be in a loop.

andyzheng0831 replied:
Will add "--retry 5 --retry-delay 3" like other curl usages in this PR.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,684 @@
+	fi
+fi
+start-fluentd
+echo "Done for the configuration for kubernetes"
```
In configure-vm.sh @zmerlynn did a neat trick setting the MOTD to show whether the startup script was still running or had finished. Can we do something similar on GCI?

andyzheng0831 replied:
This script is run by systemd service kube-master-configuration.service on master and kube-node-configuration.service on node. We can check if it is still running with "systemctl status -u kube-master-configuration.service" and "systemctl status -u kube-node-configuration.service". And we can see the log using "journalctl -u".

zmerlynn replied:
We also need all the licensing work that `configure-vm.sh` has.

andyzheng0831 replied:
The "licensing work" means https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/configure-vm.sh#l104-116?

roberthbailey commented on cluster/gce/gci/health-monitor.sh:
```diff
@@ -0,0 +1,50 @@
+# We simply kill the process when there is a failure. Another upstart job will automatically
+# restart the process.
+while [ 1 ]; do
+  if ! timeout 10 docker version > /dev/null; then
```
should the docker & kubelet checks be coupled this way (where they run serially)? In the debian image we run them in parallel.

andyzheng0831 replied:
Agree with you that we should decouple them. The checking cycle of two components may interfere with each other. I check the supervior code, but unfortunately we cannot directly use the scripts from there. Considering this health-monitor.sh is small, I prefer not to split it into two files. Instead, we add a parameter for triggering either docker or kubelet monitoring. In master.yaml and node.yaml, we will have two daemon type systemd job, each for monitoring one component.

roberthbailey commented:
@zmerlynn can you take a look as well since these scripts are closely related to configure-vm.sh?

andyzheng0831 commented:
@roberthbailey An updated version is uploaded, which addressed your comments. But we still have a couple of comments waiting for more information from @zmerlynn and @mikedanese . I am reruning e2e for at least 5 times against this new version

andyzheng0831 commented:
@zmerlynn @roberthbailey I just updated the PR by adding the /etc/motd support (see functions set-broken-motd and reset-motd) and licensing work. Since /usr/local/ is readonly in GCI, I push LICENSE file in /home/kubernetes, and display the info by /etc/motd. I tested the new version using e2e tests.

zmerlynn commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,750 @@
+# Finds the master PD device; returns it in MASTER_PD_DEVICE
+function find-master-pd {
+  MASTER_PD_DEVICE=""
+  # TODO(zmerlynn): GKE is still lagging in master-pd creation.
```
Since you forced into rebase anyways, can you remove this TODO from both sides? I've been meaning to remove it from one, but I'd rather not see it copied. :) (It's been done for a while.)

andyzheng0831 replied:
Sure, will remove it

zmerlynn commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,750 @@
+  fi
+  local -r kube_proxy_docker_tag=$(cat /home/kubernetes/kube-docker-files/kube-proxy.docker_tag)
+  local api_servers="--master=https://${KUBERNETES_MASTER_NAME}"
+  sed -i -e "s@{{kubeconfig}}@${kubeconfig}@g" ${src_file}
```
FWIW, and we can resolve this as a follow-on, it looks like this won't work on OS X (and probably hasn't ever) (BSD `sed` has a different syntax for `-i`). We should just start searching for gnu *everything* like the build script searches for gnu `tar`.

zmerlynn replied:
Ignore me, node side.

zmerlynn commented:
LGTM. I left a nit but you can fix it or not.

andyzheng0831 commented:
@zmerlynn @roberthbailey The PR is rebased. I am running e2e on GCI again to make sure no new problem introduced.

roberthbailey commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -0,0 +1,749 @@
+    cp "${KUBE_HOME}/kube-manifests/kubernetes/kube-registry-proxy.yaml" /etc/kubernetes/manifests
+	fi
+fi
+reset-motd
```
This should be the last thing that we do. Why do you configure fluentd after putting up the message?

andyzheng0831 replied:
Probably a paste error. Will update it.

roberthbailey commented on cluster/gce/gci/health-monitor.sh:
```diff
@@ -0,0 +1,80 @@
+# automatically restart the process.
+function docker_monitoring {
+  while [ 1 ]; do
+    if ! timeout 10 docker version > /dev/null; then
```
https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/supervisor/docker-checker.sh has switched to using `docker ps` instead of `docker version`. We should make this consistent. 

/cc @dchen1107

andyzheng0831 replied:
Will update it

roberthbailey commented on cluster/gce/gci/health-monitor.sh:
```diff
@@ -0,0 +1,80 @@
+  local -r max_seconds=10
+  while [ 1 ]; do
+    if ! curl --insecure -m "${max_seconds}" -f -s https://127.0.0.1:${KUBELET_PORT:-10250}/healthz > /dev/null; then
+      echo "Kubelet is unhealthy!"
```
https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/supervisor/kubelet-checker.sh adds an extra curl here that isn't piped to `/dev/null` before killing the kubelet. For debugging, it would be nice to know what healthz was reporting prior to being killed. In container-vm, that log would end up in a supervisor log location, where can we put it on GCI?

andyzheng0831 replied:
The extra curl uses http, but the first curl uses https. Is this intentionally or a typo?

andyzheng0831 replied:
@roberthbailey @dchen1107 I just verified that the two curls in https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/supervisor/kubelet-checker.sh produce different things. The second one is incorrect. I can update it in this PR, or leave it to others to fix.

roberthbailey replied:
My guess is a typo. Let's fix it in this PR (so that we don't add a 1 line PR to the submit queue).

andyzheng0831 replied:
Fixed

andyzheng0831 commented:
Let me also backport #25504 into this PR. In anyway I will need to merge it from trusty code to gci.

andyzheng0831 commented:
Please check the latest version. I already ran e2e on GCI to verify it.

andyzheng0831 commented:
Rebase done

roberthbailey commented:
@lavalamp -- Adding the e2e-not-required label since the only part of this PR that would be tested by the submit queue in the 1 line change to `kubelet-checker.sh` which is low risk.

lavalamp commented:
GKE cluster start up broke... this is the only thing that changed the start up path or that looks remotely suspicious. Reverting, sorry.

wonderfly commented:
Looks like there is no plan to cherry pick this to `release-1.2`? I am asking because the support of `OS_DISTRIBUTION=gci` in `release-1.2` is broken in the sense that it doesn't let users choose [`MASTER_IMAGE_PROJECT`](https://github.com/kubernetes/kubernetes/blob/release-1.2/cluster/gce/util.sh#L31) so all GCI release tests (that grab GCI images from a staging project) will fail if they set `OS_DISTRIBUTION=gci` (as what's happening right now).

I'd like to send a PR directly to the `release-1.2` branch to fix that, but that doesn't seem to be a recommended practice...

roberthbailey commented:
> Looks like there is no plan to cherry pick this to `release-1.2`?

Not that I know of. I think @andyzheng0831's plan was to keep using trusty, since it exists on that release branch.
